src/prods/p10.py

```diff
@@ -0,0 +1,155 @@
+        hang = Node(NodeAttrs('v', x, y, hanging=False), hdl)
+        graph.add_node(hang)
+        # recreate edges with proper boundry property
+        new_attr = EdgeAttrs('e', is_boundry)
+        edge = Edge(vertices[0].handle, hdl, new_attr)
+        graph.add_edge(edge)
+        new_attr = EdgeAttrs('e', is_boundry)
+        edge = Edge(vertices[1].handle, hdl, new_attr)
+        graph.add_edge(edge)
+
+        # skip first pair because it already exists
+        idx_nodes = [1, 4, 2, 3, 0]
+        nodes_shuffled = [vertices[idx] for idx in idx_nodes]
+
+        # hanging node is already created, should be on boundry list
+        new_boundry = [hang]
+        for node_a, node_b in it.pairwise(nodes_shuffled):
+            x = (node_a.attrs.x + node_b.attrs.x) / 2
+            y = (node_a.attrs.y + node_b.attrs.y) / 2
+            edge_attrs = graph.edge_attrs((node_a.handle, node_b.handle))
+            hanging = not edge_attrs.value
+            new_node = Node(NodeAttrs('v', x, y, hanging))
+            new_boundry.append(new_node)
+
+            graph.remove_edge(node_a.handle, node_b.handle)
+            graph.add_node(new_node)
+
+            new_attrs = EdgeAttrs('e', edge_attrs.value)
+            edge = Edge(u=node_a.handle, v=new_node.handle, attrs=new_attrs)
+            graph.add_edge(edge)
+            new_attrs = EdgeAttrs('e', edge_attrs.value)
+            edge = Edge(u=new_node.handle, v=node_b.handle, attrs=new_attrs)
+            graph.add_edge(edge)
+
+        graph.remove_node(node_p.handle)
+
+        x = node_p.attrs.x
+        y = node_p.attrs.y
+        central = Node(NodeAttrs('v', x, y, hanging=False), node_p.handle)
+        graph.add_node(central)
+
+        for new_node in new_boundry:
+            attr = EdgeAttrs('e', False)
+            edge = Edge(new_node.handle, central.handle, attr)
+            graph.add_edge(edge)
+
+        # connect new nodes on the boundry with inner nodes
+        inner_nodes = []
+        for vert, (node_a, node_b) in zip([vertices[0]] + nodes_shuffled, it.pairwise([new_boundry[-1]] + new_boundry)):
+            x = (vert.attrs.x + node_a.attrs.x + node_b.attrs.x + node_p.attrs.x) / 4
+            y = (vert.attrs.y + node_a.attrs.y + node_b.attrs.y + node_p.attrs.y) / 4
+            inner_node = Node(NodeAttrs('q', x, y, hanging=False))
+            inner_nodes.append(inner_node)
+            graph.add_node(inner_node)
+            attr = EdgeAttrs('q', False)
+            edge = Edge(node_a.handle, inner_node.handle, attr)
+            graph.add_edge(edge)
+            attr = EdgeAttrs('q', False)
+            edge = Edge(node_b.handle, inner_node.handle, attr)
+            graph.add_edge(edge)
+            attr = EdgeAttrs('q', False)
+            edge = Edge(node_p.handle, inner_node.handle, attr)
+            graph.add_edge(edge)
+            attr = EdgeAttrs('q', False)
+            edge = Edge(vert.handle, inner_node.handle, attr)
+            graph.add_edge(edge)
```

Comment:
This could work, but it actually breaks some invariants `add_q_hyperedge` method introduced: 

1. All edges that comprise the Q hyperedge have the same handle, here each edge has its own EdgeAttrs, thus its own handle.

This could be solved if you had just single `attr` object, and pass it everywhere or just make use of `add_q_hyperedge` method.